testcases/kernel/syscalls/fanotify/fanotify09.c

```diff
@@ -0,0 +1,236 @@
+ * any, provided herein do not apply to combinations of this program with
+ * other software, or any other product whatsoever.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Started by Amir Goldstein <amir73il@gmail.com>
+ *
+ * DESCRIPTION
+ *     Check that fanotify handles events on children correctly when
+ *     both inode and mountpoint marks exist.
+ *
+ * This is a regression test for commit 54a307ba8d3c:
+ *
+ *      fanotify: fix logic of events on child
+ */
+#define _GNU_SOURCE
+#include "config.h"
+
+#include <stdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <errno.h>
+#include <string.h>
+#include <sys/mount.h>
+#include <sys/syscall.h>
+#include "tst_test.h"
+#include "fanotify.h"
+
+#if defined(HAVE_SYS_FANOTIFY_H)
+#include <sys/fanotify.h>
+
+#define EVENT_MAX 1024
+/* size of the event structure, not counting name */
+#define EVENT_SIZE  (sizeof (struct fanotify_event_metadata))
+/* reasonable guess as to size of 1024 events */
+#define EVENT_BUF_LEN        (EVENT_MAX * EVENT_SIZE)
+
+static void setup(void);
+static void cleanup(void);
+
+#define NUM_GROUPS 3
+
+#define BUF_SIZE 256
+static char fname[BUF_SIZE];
+static int fd_notify[NUM_GROUPS];
+
+static char event_buf[EVENT_BUF_LEN];
+
+#define MOUNT_NAME "mntpoint"
+static int mount_created;
+
+static void create_fanotify_groups(void)
+{
+	unsigned int i, onchild;
+	int ret;
+
+	for (i = 0; i < NUM_GROUPS; i++) {
+		fd_notify[i] = SAFE_FANOTIFY_INIT(FAN_CLASS_NOTIF |
+						FAN_NONBLOCK,
+						O_RDONLY);
+
+		/* Add mount mark for each group without MODIFY event */
+		ret = fanotify_mark(fd_notify[i],
+				    FAN_MARK_ADD | FAN_MARK_MOUNT,
+				    FAN_CLOSE_NOWRITE,
+				    AT_FDCWD, ".");
+		if (ret < 0) {
+			tst_brk(TBROK | TERRNO,
+				 "fanotify_mark(%d, FAN_MARK_ADD | "
+				 "FAN_MARK_MOUNT, FAN_MODIFY, AT_FDCWD,"
+				 " '.') failed", fd_notify[i]);
+		}
+		/*
+		 * Add inode mark on parent for each group with MODIFY
+		 * event, but only one group requests events on child.
+		 * The one mark with FAN_EVENT_ON_CHILD is needed for
+		 * setting the DCACHE_FSNOTIFY_PARENT_WATCHED dentry
+		 * flag.
+		 */
+		onchild = (i == 0) ? FAN_EVENT_ON_CHILD : 0;
+		ret = fanotify_mark(fd_notify[i],
+				    FAN_MARK_ADD,
+				    FAN_MODIFY | onchild, AT_FDCWD, ".");
+		if (ret < 0) {
+			tst_brk(TBROK | TERRNO,
+				 "fanotify_mark(%d, FAN_MARK_ADD, "
+				 "FAN_MODIFY%s, AT_FDCWD, '.') failed",
+				 fd_notify[i],
+				 onchild ? " | FAN_EVENT_ON_CHILD" : "");
+		}
+	}
+}
+
+static void cleanup_fanotify_groups(void)
+{
+	unsigned int i;
+
+	for (i = 0; i < NUM_GROUPS; i++) {
+		if (fd_notify[i] && fd_notify[i] != -1) {
+			if (close(fd_notify[i]) == -1)
+				tst_res(TWARN, "close(%d) failed",
+					 fd_notify[i]);
+			fd_notify[i] = 0;
+		}
```

Comment:
We can use SAFE_CLOSE() even in the test cleanup() in the new library, hence the whole inner loop could be simplified just to SAFE_CLOSE(fd_notify[i]);